README
Managed conferencing solution using 
Asterisk, Web-MeetMe and a database-driven scheduler
by Dan Austin <dan_austin@fitawi.com>
================================================================

** Background
	It has been suggested that conferencing in general and 
MeetMe specifically represent the killer app to drive Asterisk
integrations.  While MeetMe does provide a solid foundation for
audio conferencing, there is a need for a management interface
that allows system administrators, conference administrators,
and system users to schedule, monitor and manage conferences.

** Security concerns
        It is strongly recommended that register_globals be
turned off.  The Web-MeetMe UI no longer needs this feature
turned on, and if it is turned on the system may be vulnerable
to XSS.

** Using Web-MeetMe with Asterisk 1.6 or newer
	Starting with Asterisk 1.6, the logic to enforce
conference start times, end times and resource limits have
been moved into app_meetme.  This means that this package
no longer needs app_cbmysql, however an updated verion
is included for sites that already use WMM and cannot
convert to using RealTime.  It is highly recommended that
new installations use RealTime and not app_cbmysql.

** The tools

Web-MeetMe
	Web-MeetMe is a collection of PHP pages that leverage
the database for scheduling and the Asterisk Manager interface
to monitor and control active conferences.

	The scheduler currently supports these features:
		+ Conference number conflict avoidance
		+ Recurring conferences 
		  * Daily, weekly or bi-weekly 
		+ Enforces a user password if a admin password 
		  is set
		+ Seperate views for past, current and future
		  conferences
		+ Editing future conferences
		  * Includes a series of conferences
		+ Deleting past conferences
		+ CDR-like view of past conferences
		  * Includes a link to conference recordings
		+ One click features to-
		  * Extend an active conference (time)
		  * Kick out all participants
		  * Out-call to include participants 
		+ Optional early join features
		  * fuzzystart allows a caller to enter a
		    room early (in seconds, set in cbmysql.conf)
		  * earlyalert notifies a call that their 
		    conference has not yet started, if they
		    enter a valid conference number and the
		    conference is scheduled to start soon.
		    (in seconds, set in cbmysql.conf)

** Installation
* Dependencies
	-  Asterisk 1.6 or later
		!! Version 1.6.2.7 is the first suitable version
		!! Version 1.8.0 is recommended for full functionality
		!! res_odbc
		!! app_meetme configured to use RealTime
		!! cdr_adaptive_odbc is recommended for CDR functionality
	-  Database with ODBC driver
	-  PHP 4.3 or later
		!! Reports depend on mod_gd and GD-2.0.28
		!! or later
		!! PEAR::DB (php-pear)

* Tweak meetme.conf
	The default meetme.conf from Asterisk 1.6.2 or later has
the basics in place.  One exception is that logmembercount should
be set to 'no', as that feature of Realtime MeetMe conference is
not used by WMM.

* Configure ODBC
	Configure /etc/odbc.ini and add a section similar to this:

[MySQL-Asterisk]
Description     = ODBC for MySQL
Driver          = MySQL
Server          = 127.0.0.1
Database        = meetme
Port            = 3306
Socket          = /tmp/mysql.sock
Option          =
Stmt            =
Trace           = yes
TraceFile       = /tmp/odbc.log


	Configure /etc/asterisk/res_odbc.conf and add a section similar
to this:

[meetme]
dsn => MySQL-Asterisk
username => <db-username>
password => <db-password>
pre-connect => yes

	Add the following line to /etc/asterisk/extconfig.conf:

meetme => odbc,meetme,booking

	For CDR logs configure /etc/asterisk/cdr_adaptive_odbc.conf:

[wmm]
connection=meetme	;Note that this matches res_odbc.conf
table=cdr

	 In /etc/asterisk/cdr.conf confirm that endbeforehexten is set to 'no'

* Setup the database
	Create a table called booking in your database.  The
table is described in ./cbmysql/db-tables-v7.txt

	You can upgrade your database from an earlier install
by using this command:

mysql < ./cbmysql/upgrade-db.txt

	There is also a use the rollback-db.txt file to
revert the database to be compatible with WMM 3.X if you
need to downgrade.

	If you will be using a SQL database to store user 
accounts instead of using an LDAP service, you can use this 
command to setup the initial administrator account to logon to 
WMM:
mysql < cbmysql/db-admin-user-create.txt


* Install Web-MeetMe
	Download the latest Web-MeetMe tgz file from
http://www.sf.net/projects/web-meetme and extract it to your webservers
document root.

	Modify the ../web-meetme/lib/defines.php to match you
database settings.

	Copy ../web-meetme/phpagi/phpagi.example.conf to 
/etc/asterisk and modify it to match the settings in your
manager.conf

********************************************************* 
You must have an account in /etc/asterisk/manager.conf
similar to this:

[MeetMe]
secret = {pick-a-password}
read = call
write = command,originate

********************************************************* 


	If you intend to use the authentication functions,
it is strongly recommended that you use SSL and force all
conenctions to the Web-MeetMe pages to use HTTPS.

* LDAP or Active Directory integration
	Edit ./lib/defines and set the AUTH_TYPE to adLDAP.
Set the ADMIN_GROUP to appropriate list, the default is
"Domain Admins", but you might not want the network folks
controlling your telephony systems.  Optionally you can
chanege the browser session timeout.

	Edit ./lib/adLDAP and replace all instances of
"yourdomain" and "yourserver" with the values for your
network.

* Usage
	If all of the steps have been followed, you should be
able to open your browser, connect to the Web-MeetMe page and
schedule a conference.  A new conference start time defaults
to the time it is scheduled, so modify it if needed.  Any
required fields that are left blank will be auto-generated.
Note the conference room number and password.

	Dial the number you have assigned for conferencing, and
try the conference you have just added.  If the time and password
are correct, you will be joined to the conference.

	You can now use the Web-Meetme monitor function to manage
your conference.

** Credits
Areski for his fantastic Web-MeetMe gui which proved extremely 
easy to extend.

Mike Clark for UI enhancements and the database features to host
user accounts.


And special thanks to everyone in the Asterisk community.



Changelog
------------------------------------------------------------------------
r260 | dan_austin | 2010-02-17 11:45:15 -0800 (Wed, 17 Feb 2010) | 1 line

More unassigned variables cleaned up, fix authentication seeming to require two logons.  Make [Un]Check_All work on Firefox.  Drop unused columns from db scripts and disable verbose debugging by default
------------------------------------------------------------------------
r263 | dan_austin | 2010-02-22 10:51:33 -0800 (Mon, 22 Feb 2010) | 1 line

Really fix [Un]Check_All for Firefox
------------------------------------------------------------------------
r270 | dan_austin | 2010-07-07 13:24:32 -0700 (Wed, 07 Jul 2010) | 1 line

Russian translation
------------------------------------------------------------------------
r273 | dan_austin | 2010-10-20 15:29:22 -0700 (Wed, 20 Oct 2010) | 1 line

Add minutes used to Reports.  Alter how reports detect used conferences to reduce DB load.  Better support for 'Requests Floor' events. Alter sample extensions.conf to eliminate empty CDR records.
------------------------------------------------------------------------
r274 | dan_austin | 2010-10-20 15:35:39 -0700 (Wed, 20 Oct 2010) | 1 line

Update db scripts.
------------------------------------------------------------------------
r277 | dan_austin | 2011-03-08 10:11:31 -0800 (Tue, 08 Mar 2011) | 1 line

Update readme and remove pagination from monitor code
------------------------------------------------------------------------
r279 | dan_austin | 2011-03-16 09:48:39 -0700 (Wed, 16 Mar 2011) | 1 line

Additional logging and table configuration options in app_cbmysql
------------------------------------------------------------------------
r281 | dan_austin | 2011-10-20 09:31:55 -0700 (Thu, 20 Oct 2011) | 1 line

Merge to trunk
------------------------------------------------------------------------
r284 | dan_austin | 2011-11-28 12:16:32 -0800 (Mon, 28 Nov 2011) | 1 line

Unify conf_actions and clean-up monitor code, restore missing recording filename support in the database
------------------------------------------------------------------------
r286 | dan_austin | 2011-12-07 13:30:53 -0800 (Wed, 07 Dec 2011) | 1 line

Fix play.php to use configured recording file path.
------------------------------------------------------------------------


lib/defines.php
<?php

include("../locale.php");

//To debug uncomment the following
//ini_set('display_errors', TRUE);
//error_reporting(E_ALL ^ E_NOTICE);

//The AJAX features require a consistent URL
//Make sure to set WEBROOT to the exact URL
//Users will use to access WMM
define ("WEBROOT", "https://dev-asterisk.phoenix.com/web-meetme/");
define ("FSROOT", "/var/www/html/web-meetme/");
define ("LIBDIR", FSROOT."lib/");

//GUI title
define("GUI_TITLE", "Web-MeetMe");
define("GUI_ICON", "asterisk.gif");
define("GUI_VER", "4.0.5");
define("GUI_SRC", "http://sourceforge.net/projects/web-meetme/");

//Email Variables - Support contacts, Call in numbers and other spit and polish for the about page.
define("LOCAL_SUPPORT", "Support Department");
define("LOCAL_PHONE", "800-123-4567");
define("PHONENUM", "800-858-4032");
define("PBX_ICON", "asterisk.gif");

//Maximum concurrent caller limit
//define("MAX_CALLER_LIMT", "9");

// Conference monitor options
define("MON_REFRESH", "5");

// Recording path
define("RECORDING_PATH", "/var/lib/asterisk/sounds/conf-recordings/");

// THIS VARIABLE DEFINE THE COLOR OF THE HEAD TABLE
$FG_TABLE_HEAD_COLOR = "#D1D9E7";
$FG_TABLE_EXTERN_COLOR = "#7F99CC"; //#CC0033 (Rouge)
$FG_TABLE_INTERN_COLOR = "#EDF3FF"; //#FFEAFF (Rose)

$FG_TABLE_ALTERNATE_ROW_COLOR[] = "#1FFFFF";
$FG_TABLE_ALTERNATE_ROW_COLOR[] = "#F2F8FF";

// THIS VARIABLE DEFINE THE COLOR OF THE ADMIN ROW
$FG_TABLE_ROW_COLOR_ADMIN = "#FCCDCA";


// Comment out the following lines to disable authentication
define ("AUTH_TYPE", "adLDAP"); // adLDAP or sqldb 
define ("ADMIN_GROUP", "Domain Admins");
define ("AUTH_TIMEOUT", "3");	//Hours
include (LIBDIR.AUTH_TYPE.".php");


//Database tables
define ("DB_TABLECDR", "cdr");
define ("DB_TABLESCHED", "booking");
define ("DB_TABLEUSERS","user");

define ("SERVER_TZ", "PST/PDT");
//define ("USE_24H", "YES");

//Outcall defaults
define ("CHAN_TYPE", "Local"); //Use Local to let dialplan decide which chan
define ("OUT_CONTEXT", "default"); //Select a context to place the call from
define ("OUT_PEER", ""); // Use this if not using CHAN_TYPE Local
define ("OUT_CALL_CID", "Parlez <1996>"); // Caller ID for Invites

//Standard flags for Users and Admins
define ("SAFLAGS", "aAs");
define ("SUFLAGS", "s");
$Mod_Options = array(array(_("Announce"), "I"), array(_("Record"), "r"));
$User_Options = array(array(_("Announce"), "I"), array(_("Listen Only"), "m"), array(_("Wait for Leader"), "w"));

//Require conference PIN (passwords)
define ("PASSWORD_OPTION", "NO");

//Change conference End Time on a 'End Now' click
//define ("FORCE_END", "YES");

//Mailer type: 
// CLIENT to use mailto: and default user mail client
// SERVER to use the server's mailer
define ("MAILER", "CLIENT");
include ("email_body.php");

//Avatar definitions
$icons_list['0'] = "./images/icons/Darth Vader.GIF"; 
$icons_list['1'] = "./images/icons/Anakin.GIF";
$icons_list['2'] = "./images/icons/Scout Trooper.GIF";
$icons_list['3'] = "./images/icons/Hoth Soldier.GIF";
$icons_list['4'] = "./images/icons/Major Matt Mason.GIF";
$icons_list['5'] = "./images/icons/Landspeeder Ben.GIF";
$icons_list['6'] = "./images/icons/Obi Wan With Hood.GIF";
$icons_list['7'] = "./images/icons/Wedge.GIF";
$icons_list['8'] = "./images/icons/Rebel Tech.GIF";
$icons_list['9'] = "./images/icons/Larry the Lobster.GIF";



$months = array(_("January"),_("February"),_("March"),_("April"),_("May"),_("June"),_("July"),_("August"),_("September"),_("October"),_("November"),_("December"));

$days = array("31","28","31","30","31","30","31","31","30","31","30","31");

$recurLabel = array(_("Daily"), _("Weekly"), _("Bi-weekly"));
$recurInterval = array("86400", "604800", "1209600");
$recurPeriod = array("14", "26", "13");

function contact(){
	?>      
	<table width="90%">
	<tr><td couluns=2 ><br><h1><?php print _("Support Team"); ?> ...</h1></td></tr>
	<tr>
	<td><em><strong>Contact: </strong></em><?php print _("Contact"); ?></td>
	<td><em><strong>Phone: </strong></em><?php _("Phone"); ?></td>
	</tr>

	<tr><td span=2 ><br><h2><?php print _("Developer Team"); ?> ...</h2></td></tr>
	<tr>
    	<td><a href="mailto:<?=$myemail?>">Arezqui Bela&iuml;d
      	</a> : areski (no@spam) gmail (dot) com</td>
    	<td><a href="mailto:dan_austin@phoenix.com">Dan Austin
      	</a></td>
  	</tr>
	<tr>
	<td><em><strong>Last update: </strong></em><?php  _("Last update"); ?></td>
	<td><em><strong>Developer Website: </strong></em><a href="<?php _("Developer Website"); ?>">Web-MeetMe</a></td>
	</tr>
	<tr>
	<td><br><img src="images/<?=PBX_ICON?>"></td>
	<tr>
	<td><h2>User details ...</h2></td>
	</tr>
	<tr>
	<td> <?php print _("Currently logged on as"); ?> <?php echo $_SESSION['userid']?> a <?php echo $_SESSION['privilege']?> <?php if (isset ($_SESSION['groups'])) print _("and a member of"); ?> <?php if (isset ($_SESSION['groups'])) echo $_SESSION['groups']?></td>
	</tr>
	</table>
<br><br>
	<?php
}

?>
